File: crates/neon_cluster/src/controllers/pageserver/service.rs
use crate::util::errors::{Error, Result, StdError};

use k8s_openapi::api::core::v1::{Service, ServicePort, ServiceSpec};
use k8s_openapi::apimachinery::pkg::apis::meta::v1::{ObjectMeta, OwnerReference};

use kube::{
    api::{Patch, PatchParams, PostParams},
    Api, Client,
};

use std::collections::BTreeMap;
use tracing::info;

pub async fn reconcile_headless_service(
    client: &Client,
    namespace: &str,
    name: &str,
    cluster_name: &str,
    oref: &OwnerReference,
) -> Result<()> {
    let services: Api<Service> = Api::namespaced(client.clone(), namespace);
    let desired_service = create_desired_headless_service(namespace, name, cluster_name, oref);

    match services.get(name).await {
        Ok(existing) => {
            if service_needs_update(&existing, &desired_service) {
                info!("Updating Headless Service '{}'", name);
                services
                    .patch(
                        name,
                        &PatchParams::apply("kube-rs-controller").force(),
                        &Patch::Apply(&desired_service),
                    )
                    .await
                    .map_err(|e| Error::StdError(StdError::KubeError(e)))?;
            } else {
                info!("Headless Service '{}' is up to date", name);
            }
        }
        Err(kube::Error::Api(api_err)) if api_err.code == 404 => {
            info!("Creating Headless Service '{}'", name);
            services
                .create(&PostParams::default(), &desired_service)
                .await
                .map_err(|e| Error::StdError(StdError::KubeError(e)))?;
        }
        Err(e) => return Err(Error::StdError(StdError::KubeError(e))),
    }

    Ok(())
}


fn create_desired_headless_service(
    namespace: &str,
    name: &str,
    cluster_name: &str,
    oref: &OwnerReference,
) -> Service {
    let mut labels = BTreeMap::new();
    labels.insert(
        "app.kubernetes.io/name".to_string(),
        format!("pageserver-{}", cluster_name),
    );
    labels.insert(
        "app.kubernetes.io/component".to_string(),
        "pageserver".to_string(),
    );

    Service {
        metadata: ObjectMeta {
            name: Some(name.to_string()),
            namespace: Some(namespace.to_string()),
            labels: Some(labels.clone()),
            owner_references: Some(vec![oref.clone()]),
            ..Default::default()
        },
        spec: Some(ServiceSpec {
            cluster_ip: Some("None".to_string()), // Headless service
            selector: Some(labels),
            ports: Some(vec![
                ServicePort {
                    name: Some("pageserver-pg".to_string()),
                    port: 6400,
                    target_port: Some(k8s_openapi::apimachinery::pkg::util::intstr::IntOrString::Int(
                        6400,
                    )),
                    protocol: Some("TCP".to_string()),
                    ..Default::default()
                },
                ServicePort {
                    name: Some("pageserver-http".to_string()),
                    port: 9898,
                    target_port: Some(k8s_openapi::apimachinery::pkg::util::intstr::IntOrString::Int(
                        9898,
                    )),
                    protocol: Some("TCP".to_string()),
                    ..Default::default()
                },
            ]),
            ..Default::default()
        }),
        ..Default::default()
    }
}


pub fn service_needs_update(existing: &Service, desired: &Service) -> bool {
    let existing_spec = existing.spec.as_ref().unwrap();
    let desired_spec = desired.spec.as_ref().unwrap();

    existing_spec.cluster_ip != desired_spec.cluster_ip
        || existing_spec.selector != desired_spec.selector
        || existing_spec.ports != desired_spec.ports
}

File: specs/compute/service.go
package compute

import (
	"fmt"

	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/util/intstr"

	neonv1alpha1 "oltp.molnett.org/neon-operator/api/v1alpha1"
)

type ServiceConfig struct {
	Suffix    string
	Component string
	PortName  string
	Port      int32
}

func createService(branch *neonv1alpha1.Branch, project *neonv1alpha1.Project, config ServiceConfig) *corev1.Service {
	serviceName := fmt.Sprintf("%s-%s", branch.Name, config.Suffix)
	labels := map[string]string{
		"molnett.org/cluster":   project.Spec.ClusterName,
		"molnett.org/component": config.Component,
		"molnett.org/branch":    branch.Name,
		"neon.timeline_id":      branch.Spec.TimelineID,
		"neon.tenant_id":        project.Spec.TenantID,
	}

	return &corev1.Service{
		TypeMeta: metav1.TypeMeta{
			APIVersion: "v1",
			Kind:       "Service",
		},
		ObjectMeta: metav1.ObjectMeta{
			Name:      serviceName,
			Namespace: branch.Namespace,
			Labels:    labels,
		},
		Spec: corev1.ServiceSpec{
			Selector: map[string]string{
				"app": fmt.Sprintf("%s-compute-node", branch.Name),
			},
			Ports: []corev1.ServicePort{
				{
					Name:       config.PortName,
					Port:       config.Port,
					Protocol:   corev1.ProtocolTCP,
					TargetPort: intstr.FromInt(int(config.Port)),
				},
			},
		},
	}
}

func AdminService(branch *neonv1alpha1.Branch, project *neonv1alpha1.Project) *corev1.Service {
	return createService(branch, project, ServiceConfig{
		Suffix:    "admin",
		Component: "compute-admin",
		PortName:  "admin",
		Port:      3080,
	})
}

func PostgresService(branch *neonv1alpha1.Branch, project *neonv1alpha1.Project) *corev1.Service {
	return createService(branch, project, ServiceConfig{
		Suffix:    "postgres",
		Component: "compute-postgres",
		PortName:  "postgres",
		Port:      55433,
	})
}
